src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -53,33 +56,65 @@ public override void Initialize (AnalysisContext context)
 					}
 				}
 			});
-			// TODO: fix reporting for generic type substitutions. This should happen not only for method invocations,
-			// but for any reference to an instantiated method or type.
-			context.RegisterOperationAction (context => {
-				var invocationOperation = (IInvocationOperation) context.Operation;
-				ProcessInvocationOperation (context, invocationOperation);
-			}, OperationKind.Invocation);
+			context.RegisterSyntaxNodeAction (context => {
+				ProcessGenericParameters (context);
+			}, SyntaxKind.GenericName);
 		}
 
-		static void ProcessInvocationOperation (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		static void ProcessGenericParameters (SyntaxNodeAnalysisContext context)
 		{
-			if (context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
+			if (context.ContainingSymbol is not null && context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute)) {
 				return;
+			}
 
-			ProcessTypeArguments (context, invocationOperation);
-		}
-
-		static void ProcessTypeArguments (OperationAnalysisContext context, IInvocationOperation invocationOperation)
-		{
-			var targetMethod = invocationOperation.TargetMethod;
-			if (targetMethod.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
-				return;
+			ImmutableArray<ITypeParameterSymbol> typeParams = default;
+			ImmutableArray<ITypeSymbol> typeArgs = default;
+			if (context.SemanticModel.GetTypeInfo (context.Node).ConvertedType is INamedTypeSymbol type) {
+				// INamedTypeSymbol inside nameof, commonly used in [ExpectedWarning] can generate diagnostics
+				// Walking the node heirarchy to check if INamedTypeSymbol is inside a nameof
+				var node = context.Node;
+				while (node != null) {
+					if (node is InvocationExpressionSyntax invocationExpression && invocationExpression.Expression is IdentifierNameSyntax ident1) {
+						if (ident1.Identifier.ValueText.Equals ("nameof"))
+							return;
+					}
+					node = node.Parent;
+				}
+				typeParams = type.TypeParameters;
+				typeArgs = type.TypeArguments;
+			} else if (context.SemanticModel.GetSymbolInfo (context.Node, context.CancellationToken).Symbol is IMethodSymbol targetMethod) {
+				// Check if the generic parameters have any constraints on this method
```

Comment:
The type arguments of this generic method could be generic parameters of the containing method, or of the containing type - this doesn't handle the type case.

I think instead of walking over the declaring member's method/type syntax to find constraints, you should be able to get the `targetMethod.TypeArguments`. I think those can be either an `INamedTypeSymbol` or an `ITypeParameterSymbol` (or an array type, but maybe we can leave that for a separate change). `ITypeParameterSymbol` can be that of the method or the type, and it has info about the constraints.

Reply:
Moved constraints outside of this to [GenericParameterValue ](https://github.com/dotnet/linker/blob/8485bfcfbc55ab2b7007e3d39903e502fee82727/src/ILLink.RoslynAnalyzer/TrimAnalysis/GenericParameterValue.cs#L22)as @vitek-karas suggested